packages/web-components/fast-foundation/src/virtual-list/virtual-list.ts

```diff
@@ -0,0 +1,799 @@
+import {
+    attr,
+    DOM,
+    html,
+    Notifier,
+    nullableNumberConverter,
+    Observable,
+    observable,
+    RepeatBehavior,
+    RepeatDirective,
+    Splice,
+    ViewTemplate,
+} from "@microsoft/fast-element";
+import { eventResize, eventScroll, Orientation } from "@microsoft/fast-web-utilities";
+import { FoundationElement } from "../foundation-element";
+import { IntersectionService } from "../utilities/intersection-service";
+import type { ResizeObserverClassDefinition } from "../utilities/resize-observer";
+
+/**
+ * Defines when the component updates its position automatically.
+ *
+ * @beta
+ */
+export type VirtualListAutoUpdateMode = "manual" | "viewport-resize" | "auto";
+
+/**
+ * Used to describe the position of an element within the list
+ */
+export interface SpanMap {
+    start: number;
+    end: number;
+    span: number;
+}
+
+/**
+ * The default item template
+ * Authors will typically want to provide a template specific to their needs
+ * as the default display isn't particularly useful
+ */
+const defaultItemTemplate: ViewTemplate<any> = html`
+    <div
+        style="
+            overflow-wrap: anywhere;
+            overflow: hidden;
+            position: absolute;
+            height:  ${(x, c) =>
```

Comment:
It looks like the `height`, `width`, and `transform` properties require this standard (or highly common) setup. Can we extract this somehow or apply it by default to the template. My concern is that the template will always need to be customized and this is a fairly complicated requirement that will be easy to get wrong.

Or is this container div always standard and we should put the template _inside_ of it, like the default template is just the `${x => JSON.stringify(x)}` part?

Reply:
This whole defaultItemTemplate actually exists only to show devs that something is rendering and to provide a generic example of how to create a template for their own particular content, I can't imagine anybody actually using it "as is".  The `JSON.stringify()` just turns the item data into a string so we're rendering something.   The reality of this component is that devs will have to come up with templates that suit their needs.  Other components derived from a `virtual-list` could be more specific, for example the virtualizing `data-grid` which would use this under the covers would use a `data-grid-row` as a default template because it "knows" it's rendering those.  The generic list component not so much, without author prompting we don't know how to display arbitrary data.

Reply:
Open to suggestions for better approaches on this.

Reply:
I understand that this template was a sample and would probably not be used in actual code. The concern I was bringing up is that a lot of the template seems boilerplate and would need to be duplicated in order to customize it. I'm looking for a way for the template to be more like:

```
        ${x => JSON.stringify(x)}
```

and the common foundation to be more like:
```
    <div
        style="
            overflow-wrap: anywhere;
            overflow: hidden;
            position: absolute;
            height:  ${(x, c) =>
            c.parent.orientation === Orientation.vertical
                ? `${c.parent.visibleItemSpans[c.index]?.span}px`
                : `100%`};
            width:  ${(x, c) =>
            c.parent.orientation === Orientation.vertical
                ? `100%`
                : `${c.parent.visibleItemSpans[c.index]?.span}px`};
            transform: ${(x, c) =>
            c.parent.orientation === Orientation.horizontal
                ? `translateX(${c.parent.visibleItemSpans[c.index]?.start}px)`
                : `translateY(${c.parent.visibleItemSpans[c.index]?.start}px)`};
        "
    >
        ${x => x}
    </div>
```

Unless I'm not thinking broadly enough about the item container template, it seems the primary decision is horizontal vs. vertical and what do to for the width and height in either case. That should be same regardless of what the actual item looks like, so looking for a way to keep that common.

Reply:
Actually, there is a lot of variation possible as to how to structure a template given the exposed properties and I think authors will want to optimize for their scenario to get the best perf.  I am thinking on how to put together a generically useful "virtualized-item" type component that might be easier to grok and use, perhaps at some minor perf cost, but I don't think that gets implemented at the list level.

Here are some sample templates from my prototyping efforts:

Data grid's row template still uses grid layout rather than transforms for placement.  Not sure if this will remain true when it's done, but the flexibility is useful:

https://github.com/microsoft/fast/blob/903c068b3f570fbb73f5540749554980d19fb320/packages/web-components/fast-foundation/src/data-grid/data-grid.template.ts#L9

In my test "gallery panel" the "gallery-group" component has a template that does some fancy selection magic and is optimized for the vertical orientation.

https://github.com/microsoft/fast/blob/903c068b3f570fbb73f5540749554980d19fb320/packages/web-components/fast-foundation/src/gallery-panel/gallery-group.template.ts#L10

The "gallery item" templates on the other hand is much simpler, optimized for horizontal placement, and because it is fixed it just uses css for that:
https://github.com/microsoft/fast/blob/903c068b3f570fbb73f5540749554980d19fb320/packages/web-components/fast-foundation/src/gallery-panel/gallery.template.ts#L9

So, while I do think there is value in a generic "virtual item" type component for ease of use I think that's something we add that isn't core to the "virtual-list" component.  The raw templating capabilities are powerful and should be exposed.

Reply:
I pulled my "virtual-list-item" concept to this pr.  It is still rough but should be at a state where we can discuss whether it's a good way to address this concern and make the ergonomics easier at the entry level.  
- The previous api's which allow custom list item generation templates is still in place for extremely custom scenarios.
- There is now a 'virtual-list-item component' that is used by the default item generation templates, there are now default templates for horizontal and vertical (for less calculations at runtime).
- 'list-item' has a "listItemContext" property which can be used to pass custom context to list children.
- The 'virtual-list-item' has no built-in template and a custom context type which includes a 'listItemTemplate' property which it uses instead (first time using `resolveTemplate()`).  
- Defining a template for a virtual list can be as simple as:

![image](https://user-images.githubusercontent.com/7649425/153889066-2c26b88c-f2cb-4dc1-87fa-784a9e54c045.png)

And then add it to your list with:

` stackh1.listItemContext = {
        listItemTemplate: listItemTemplate,
 };`

We'd also add loading optimizations at the list-item level, still experimenting with that.

@bheston  Does this help?

Reply:
I think that looks great. It makes the simple case easier to use and provides a sample for someone wanting to extend it.